helix-term/build.rs

```diff
@@ -1,9 +1,143 @@
 use helix_loader::grammar::{build_grammars, fetch_grammars};
 
+// #[cfg(windows)]
+use std::io::prelude::Write;
+use std::path::Path;
+use std::{io, path::PathBuf, process};
+
 fn main() {
     if std::env::var("HELIX_DISABLE_AUTO_GRAMMAR_BUILD").is_err() {
         fetch_grammars().expect("Failed to fetch tree-sitter grammars");
         build_grammars(Some(std::env::var("TARGET").unwrap()))
             .expect("Failed to compile tree-sitter grammars");
     }
+
+    if cfg!(target_os = "windows") {
+        let rc_exe = find_rc_exe().expect("Windows SDK is to be installed along with MSVC");
+
+        let output =
+            std::env::var("OUT_DIR").expect("Env var OUT_DIR should have been set by compiler");
+        let output_dir = PathBuf::from(output);
+
+        let rc_path = output_dir.join("resource.rc");
+        let icon_path = PathBuf::from("../contrib/helix-256p.ico");
+        write_resource_file(&rc_path, icon_path.to_str().unwrap()).unwrap();
+
+        let resource_file = PathBuf::from(&output_dir).join("resource.lib");
+        compile_with_toolkit_msvc(rc_exe, resource_file, rc_path);
+
+        println!("cargo:rustc-link-search=native={}", output_dir.display());
+        println!("cargo:rustc-link-lib=dylib=resource");
+    }
+}
+
+fn compile_with_toolkit_msvc(rc_exe: PathBuf, output: PathBuf, input: PathBuf) {
+    let mut command = process::Command::new(rc_exe);
+    let command = command.arg(format!(
+        "/I{}",
+        std::env::var("CARGO_MANIFEST_DIR").unwrap()
+    ));
+
+    let status = command
+        .arg(format!("/fo{}", output.display()))
+        .arg(format!("{}", input.display()))
+        .output()
+        .unwrap();
+
+    println!(
+        "RC Output:\n{}\n------",
+        String::from_utf8_lossy(&status.stdout)
+    );
+    println!(
+        "RC Error:\n{}\n------",
+        String::from_utf8_lossy(&status.stderr)
+    );
+}
+
+fn find_rc_exe() -> io::Result<PathBuf> {
+    let find_reg_key = process::Command::new("reg")
+        .arg("query")
+        .arg(r"HKLM\SOFTWARE\Microsoft\Windows Kits\Installed Roots")
+        .arg("/reg:32")
+        .arg("/v")
+        .arg("KitsRoot10")
+        .output()
+        .unwrap();
+
+    if find_reg_key.status.code().unwrap() != 0 {
+        return Err(io::Error::new(
+            io::ErrorKind::Other,
+            "Can not find Windows SDK",
+        ));
+    }
+
+    let lines = String::from_utf8(find_reg_key.stdout).expect("Should be able to parse the output");
```

Comment:
IMO Rfailing should never cause a build to fail so please don't panic here and instead return a result that you print in case this function fails